File: solution/1482. Minimum Number of Days to Make m Bouquets/1482.cpp
class Solution {
public:
    int minDays(vector<int>& bloomDay, int m, int k) {
        int l = 0, r = INT_MAX, mid;
        while (l < r) {
            mid = l + (r - l) / 2;
            if (checkOK(bloomDay, m, k, mid))
                r = mid;
            else
                l = mid + 1;
        }
        return l == INT_MAX ? -1 : l;
    }
    
    bool checkOK(vector<int>& bloomDay, int m, int k, int t) {
        int cnt = 0;
        for (int x : bloomDay) {
            if (x <= t) {
                if (++cnt == k) {
                    cnt = 0;
                    if (--m == 0)
                        return true;
                }
            } else {
                cnt = 0;
            }
        }
        return false;
    }
};

File: solution/1482. Minimum Number of Days to Make m Bouquets/1482.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        def check_ok(t):
            res = cnt = 0
            for x in bloomDay:
                if x <= t:
                    cnt += 1
                    if cnt >= k:
                        res += 1
                        cnt = 0
                        if res >= m:
                            return True
                else:
                    cnt = 0
            return False
        
        left, right = 0, 10**9 + 1
        while left < right:
            mid = (left + right) // 2
            if check_ok(mid):
                right = mid
            else:
                left = mid + 1
        return -1 if left == 10**9 + 1 else left
        
